parser/src/main/kotlin/tools/samt/parser/ASTPrinter.kt

```diff
@@ -0,0 +1,214 @@
+            }
+
+            is TypeAliasNode -> {
+                body(node.name)
+                body(node.type)
+                node.annotations.forEach(body)
+            }
+
+            is ServiceDeclarationNode -> {
+                body(node.name)
+                node.operations.forEach(body)
+                node.annotations.forEach(body)
+            }
+
+            is OperationParameterNode -> {
+                body(node.name)
+                body(node.type)
+                node.annotations.forEach(body)
+            }
+
+            is RequestResponseOperationNode -> {
+                body(node.name)
+                node.parameters.forEach(body)
+                if (node.returnType != null) {
+                    body(node.returnType)
+                }
+                node.raises.forEach(body)
+                node.annotations.forEach(body)
+            }
+
+            is OnewayOperationNode -> {
+                body(node.name)
+                node.parameters.forEach(body)
+                node.annotations.forEach(body)
+            }
+
+            is ProviderDeclarationNode -> {
+                body(node.name)
+                node.implements.forEach(body)
+                body(node.transport)
+            }
+
+            is ProviderImplementsNode -> {
+                body(node.serviceName)
+                node.serviceOperationNames.forEach(body)
+            }
+
+            is ProviderTransportNode -> {
+                body(node.protocolName)
+                if (node.configuration != null) {
+                    body(node.configuration)
+                }
+            }
+
+            is ConsumerDeclarationNode -> {
+                body(node.providerName)
+                node.usages.forEach(body)
+            }
+
+            is ConsumerUsesNode -> {
+                body(node.serviceName)
+                node.serviceOperationNames.forEach(body)
+            }
+
+            is AnnotationNode -> {
+                body(node.name)
+                node.arguments.forEach(body)
+            }
+
+            is CallExpressionNode -> {
+                body(node.base)
+                node.arguments.forEach(body)
+            }
+
+            is GenericSpecializationNode -> {
+                body(node.base)
+                node.arguments.forEach(body)
+            }
+
+            is OptionalDeclarationNode -> {
+                body(node.base)
+            }
+
+            is RangeExpressionNode -> {
+                body(node.left)
+                body(node.right)
+            }
+
+            is ObjectNode -> {
+                node.fields.forEach(body)
+            }
+
+            is ObjectFieldNode -> {
+                body(node.name)
+                body(node.value)
+            }
+
+            is ArrayNode -> {
+                node.values.forEach(body)
+            }
+
+            is BundleIdentifierNode -> {
+                node.components.forEach(body)
+            }
+
+            else -> {}
```

Comment:
I stronlgy prefer the `Node` class to be sealed so we can leave out this `else` block and get a compiler error if new nodes are added in the future

Reply:
The problem with including each node class is that a lot of the nodes do not have any child nodes, so all their handler blocks would just include a `do nothing here` comment. I see the advantage of getting a compiler error when adding new nodes though so I'm going to add those handlers anyway.